src/MSBuildLocator/MSBuildLocator.cs

```diff
@@ -158,29 +185,54 @@ public static void RegisterMSBuildPath(string msbuildPath)
                         return assembly;
                     }
 
-                    var assemblyName = new AssemblyName(eventArgs.Name);
-                    if (IsMSBuildAssembly(assemblyName))
+                    var assemblyName = new AssemblyName(assemblyNameString);
+                    return TryLoadAssembly(assemblyName);
+                }
+            };
+
+            AppDomain.CurrentDomain.AssemblyResolve += registeredHandler;
+#else
+            // Saving the handler in a static field so it can be unregistered later.
+            registeredHandler = (assemblyLoadContext, assemblyName) =>
+            {
+                // Assembly resolution is not thread-safe.
+                lock (loadedAssemblies)
+                {
+                    if (loadedAssemblies.TryGetValue(assemblyName.FullName, out var assembly))
                     {
-                        var targetAssembly = Path.Combine(msbuildPath, assemblyName.Name + ".dll");
-                        if (File.Exists(targetAssembly))
+                        return assembly;
+                    }         
+            
+                    return TryLoadAssembly(assemblyName);
+                }
```

Comment:
Consider extracting a method for the #if/#else cases. Or could the commonality be moved to TryLoadAssembly?